File: src/OpenAIRealtime.ts
import {
  Config,
  Data,
  Effect,
  Match,
  Queue,
  Redacted,
  Schedule,
  Stream,
  PubSub,
  Ref,
  Scope,
} from "effect";
import type { ServerEvent, BroadcastMessage } from "./Messages.js";
import { systemInstruction } from "./SystemPrompt.js";

const OPENAI_URL = "wss://api.openai.com/v1/realtime?model=gpt-realtime-mini";

const sessionUpdate = {
  type: "session.update",
  session: {
    type: "realtime",
    audio: {
      input: {
        format: { type: "audio/pcm", rate: 24000 },
        turn_detection: null,
        noise_reduction: null,
      },
    },
    instructions: systemInstruction,
    model: "gpt-realtime-mini",
    output_modalities: ["text"],
    tracing: "auto",
  },
};

class WebSocketError extends Data.TaggedError("WebSocketError")<{
  cause: unknown;
}> {}

type Connection = {
  ws: WebSocket;
  pubsub: PubSub.PubSub<BroadcastMessage>;
};

export class OpenAIRealtime extends Effect.Service<OpenAIRealtime>()(
  "OpenAIRealtime",
  {
    effect: Effect.gen(function* () {
      const apiKey = yield* Config.redacted("OPENAI_API_KEY");
      const connectionRef = yield* Ref.make<Connection | null>(null);
      const scope = yield* Scope.make();
      const connectMutex = yield* Effect.makeSemaphore(1);

      const connect = connectMutex.withPermits(1)(
        Effect.gen(function* () {
          const existing = yield* Ref.get(connectionRef);
          if (existing) return existing;

          yield* Effect.log("Connecting to OpenAI Realtime API...");

          const incomingQueue = yield* Queue.unbounded<ServerEvent>();
          const broadcastPubSub = yield* PubSub.unbounded<BroadcastMessage>();

          const connectWithRetry = Effect.async<WebSocket, WebSocketError>(
            (resume) => {
              const ws = new WebSocket(OPENAI_URL, {
                headers: { Authorization: `Bearer ${Redacted.value(apiKey)}` },
              });
              ws.addEventListener("open", () => resume(Effect.succeed(ws)));
              ws.addEventListener("error", (e) =>
                resume(Effect.fail(new WebSocketError({ cause: e })))
              );
            }
          ).pipe(
            Effect.retry(
              Schedule.exponential("1 second").pipe(
                Schedule.compose(Schedule.recurs(5)),
                Schedule.tapOutput((d) =>
                  Effect.log(`WebSocket connection failed, retrying in ${d}`)
                )
              )
            )
          );

          const ws = yield* Effect.acquireRelease(connectWithRetry, (ws) =>
            Effect.sync(() => ws.close()).pipe(
              Effect.tap(() => Queue.shutdown(incomingQueue)),
              Effect.tap(() => PubSub.shutdown(broadcastPubSub)),
              Effect.tap(() => Ref.set(connectionRef, null))
            )
          ).pipe(Scope.extend(scope));

          ws.addEventListener("message", (e) => {
            try {
              Queue.unsafeOffer(incomingQueue, JSON.parse(e.data as string));
            } catch (err) {
              console.error("Failed to parse OpenAI WebSocket message:", err);
            }
          });

          ws.send(JSON.stringify(sessionUpdate));

          yield* Effect.log("Connected to OpenAI Realtime API");

          const handleMessage = Match.type<ServerEvent>().pipe(
            Match.when({ type: "response.output_text.delta" }, (msg) =>
              PubSub.publish(broadcastPubSub, {
                type: "delta",
                responseId: msg.response_id,
                text: msg.delta,
              })
            ),
            Match.when({ type: "response.done" }, (msg) =>
              PubSub.publish(broadcastPubSub, {
                type: "complete",
                responseId: msg.response.id,
              })
            ),
            Match.when({ type: "error" }, (msg) =>
              Effect.gen(function* () {
                yield* Effect.logError(`OpenAI error: ${msg.error.message}`);
                yield* PubSub.publish(broadcastPubSub, {
                  type: "error",
                  message: msg.error.message,
                });
              })
            ),
            Match.orElse(() => Effect.void)
          );

          yield* Stream.fromQueue(incomingQueue).pipe(
            Stream.runForEach(handleMessage),
            Effect.catchAllCause((cause) =>
              Effect.logError("OpenAI message handler failed", cause)
            ),
            Effect.forkIn(scope)
          );

          const connection: Connection = { ws, pubsub: broadcastPubSub };
          yield* Ref.set(connectionRef, connection);
          return connection;
        })
      );

      const send = (msg: object) =>
        connect.pipe(
          Effect.flatMap((c) =>
            Effect.sync(() => c.ws.send(JSON.stringify(msg)))
          )
        );

      return {
        appendAudio: (base64: string) =>
          send({ type: "input_audio_buffer.append", audio: base64 }),
        commitBuffer: () => send({ type: "input_audio_buffer.commit" }),
        requestResponse: () => send({ type: "response.create" }),
        subscribe: connect.pipe(
          Effect.flatMap((c) => PubSub.subscribe(c.pubsub))
        ),
      } as const;
    }),
  }
) {}

File: go/openai_realtime.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

const openaiURL = "wss://api.openai.com/v1/realtime?model=gpt-realtime-mini"

// OpenAIRealtime manages the WebSocket connection to OpenAI's Realtime API.
type OpenAIRealtime struct {
	conn        *websocket.Conn
	broadcaster *Broadcaster
	writeCh     chan []byte // async write queue — decouples processor from socket I/O
	done        chan struct{}
	writeDone   chan struct{}

	// Response timing tracking
	timingMu        sync.Mutex
	lastRequestedAt time.Time            // when RequestResponse() was last called
	firstDeltaAt    map[string]time.Time // responseId → time of first delta
}

// NewOpenAIRealtime connects to the OpenAI Realtime API with retry logic and
// starts a background goroutine to read incoming messages.
func NewOpenAIRealtime(ctx context.Context, apiKey string, broadcaster *Broadcaster) (*OpenAIRealtime, error) {
	log.Println("Connecting to OpenAI Realtime API...")

	var conn *websocket.Conn
	var err error

	header := http.Header{}
	header.Set("Authorization", "Bearer "+apiKey)

	// Retry with exponential backoff, max 5 attempts
	backoff := time.Second
	for attempt := 0; attempt < 5; attempt++ {
		dialer := websocket.Dialer{}
		conn, _, err = dialer.DialContext(ctx, openaiURL, header)
		if err == nil {
			break
		}
		log.Printf("WebSocket connection failed (attempt %d/5), retrying in %v: %v", attempt+1, backoff, err)
		select {
		case <-time.After(backoff):
			backoff *= 2
		case <-ctx.Done():
			return nil, ctx.Err()
		}
	}
	if err != nil {
		return nil, fmt.Errorf("failed to connect to OpenAI after 5 attempts: %w", err)
	}

	// Send session configuration
	sessionUpdate := map[string]any{
		"type": "session.update",
		"session": map[string]any{
			"type": "realtime",
			"audio": map[string]any{
				"input": map[string]any{
					"format":          map[string]any{"type": "audio/pcm", "rate": 24000},
					"turn_detection":  nil,
					"noise_reduction": nil,
				},
			},
			"instructions":      systemInstruction,
			"model":             "gpt-realtime-mini",
			"output_modalities": []string{"text"},
			"tracing":           "auto",
		},
	}

	if err := conn.WriteJSON(sessionUpdate); err != nil {
		conn.Close()
		return nil, fmt.Errorf("failed to send session update: %w", err)
	}

	log.Println("Connected to OpenAI Realtime API")

	rt := &OpenAIRealtime{
		conn:         conn,
		broadcaster:  broadcaster,
		writeCh:      make(chan []byte, 256),
		done:         make(chan struct{}),
		writeDone:    make(chan struct{}),
		firstDeltaAt: make(map[string]time.Time),
	}

	// Start read and write loops
	go rt.readLoop()
	go rt.writeLoop()

	return rt, nil
}

func (rt *OpenAIRealtime) readLoop() {
	defer close(rt.done)

	for {
		_, message, err := rt.conn.ReadMessage()
		if err != nil {
			if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				log.Println("OpenAI WebSocket closed normally")
			} else {
				log.Printf("OpenAI WebSocket read error: %v", err)
			}
			return
		}

		var event ServerEvent
		if err := json.Unmarshal(message, &event); err != nil {
			log.Printf("Failed to parse OpenAI WebSocket message: %v", err)
			continue
		}

		switch event.Type {
		case "response.output_text.delta":
			rt.trackFirstDelta(event.ResponseID)
			rt.broadcaster.Publish(BroadcastMessage{
				Type:       "delta",
				ResponseID: event.ResponseID,
				Text:       event.Delta,
			})
		case "response.done":
			if event.Response != nil {
				rt.trackResponseDone(event.Response.ID)
				rt.broadcaster.Publish(BroadcastMessage{
					Type:       "complete",
					ResponseID: event.Response.ID,
				})
			}
		case "error":
			if event.Error != nil {
				log.Printf("OpenAI error: %s", event.Error.Message)
				rt.broadcaster.Publish(BroadcastMessage{
					Type:    "error",
					Message: event.Error.Message,
				})
			}
		}
	}
}

// writeLoop drains the write channel and sends messages to the WebSocket.
// This decouples the audio processor from socket I/O latency.
func (rt *OpenAIRealtime) writeLoop() {
	defer close(rt.writeDone)
	for msg := range rt.writeCh {
		if err := rt.conn.WriteMessage(websocket.TextMessage, msg); err != nil {
			log.Printf("OpenAI WebSocket write error: %v", err)
			return
		}
	}
}

// Pre-serialized JSON fragments to avoid map allocation + json.Marshal per chunk.
var (
	appendAudioPrefix = []byte(`{"type":"input_audio_buffer.append","audio":"`)
	appendAudioSuffix = []byte(`"}`)
	commitBufferMsg   = []byte(`{"type":"input_audio_buffer.commit"}`)
	responseCreateMsg = []byte(`{"type":"response.create"}`)
)

// appendBuf is reused across AppendAudio calls to avoid allocations.
// Safe because only the audio processor goroutine calls AppendAudio.
var appendBuf []byte

func (rt *OpenAIRealtime) AppendAudio(base64Audio string) {
	needed := len(appendAudioPrefix) + len(base64Audio) + len(appendAudioSuffix)
	if cap(appendBuf) < needed {
		appendBuf = make([]byte, 0, needed*2)
	}
	appendBuf = appendBuf[:0]
	appendBuf = append(appendBuf, appendAudioPrefix...)
	appendBuf = append(appendBuf, base64Audio...)
	appendBuf = append(appendBuf, appendAudioSuffix...)

	// Copy into a new slice for the channel since appendBuf is reused
	msg := make([]byte, len(appendBuf))
	copy(msg, appendBuf)
	rt.writeCh <- msg
}

func (rt *OpenAIRealtime) CommitBuffer() {
	rt.writeCh <- commitBufferMsg
}

func (rt *OpenAIRealtime) RequestResponse() {
	rt.timingMu.Lock()
	rt.lastRequestedAt = time.Now()
	rt.timingMu.Unlock()
	rt.writeCh <- responseCreateMsg
}

func (rt *OpenAIRealtime) trackFirstDelta(responseID string) {
	rt.timingMu.Lock()
	defer rt.timingMu.Unlock()

	if _, seen := rt.firstDeltaAt[responseID]; seen {
		return
	}

	now := time.Now()
	rt.firstDeltaAt[responseID] = now

	if !rt.lastRequestedAt.IsZero() {
		latency := now.Sub(rt.lastRequestedAt)
		log.Printf("[KPI] response_latency: %dms (response %s)", latency.Milliseconds(), responseID)
	}
}

func (rt *OpenAIRealtime) trackResponseDone(responseID string) {
	rt.timingMu.Lock()
	defer rt.timingMu.Unlock()

	if start, ok := rt.firstDeltaAt[responseID]; ok {
		totalTime := time.Since(start)
		log.Printf("[KPI] response_total_time: %dms (response %s)", totalTime.Milliseconds(), responseID)
		delete(rt.firstDeltaAt, responseID)
	}
}

func (rt *OpenAIRealtime) Subscribe() (<-chan BroadcastMessage, func()) {
	return rt.broadcaster.Subscribe()
}

func (rt *OpenAIRealtime) Close() {
	close(rt.writeCh)    // signal write loop to stop
	<-rt.writeDone       // wait for pending writes to flush
	rt.conn.Close()
	<-rt.done            // wait for read loop to finish
}
